src/plugins/factoids/storage.internal.js

```diff
@@ -0,0 +1,171 @@
+/* eslint-disable no-underscore-dangle */
+const fs = require('fs');
+const path = require('path');
+const { promisify, inspect } = require('util');
+
+const readFile = promisify(fs.readFile);
+const writeFile = promisify(fs.writeFile);
+
+const FILE_PATH = path.join(__dirname, 'factoids.json');
+
+function toKey(key) {
+  return (key && key.toLowerCase()) || null;
+}
+
+class Store {
+  constructor() {
+    this._items = new Map();
+    this._loaded = false;
+    this._readPromise = null;
+    this._writePromise = null;
+    this._dirty = false;
+  }
+
+  _replaceAll(items) {
+    this._items = items;
+  }
+
+  set(key, entry) {
+    this._items.set(key.toLowerCase(), entry);
+    this._dirty = true;
+    return this;
+  }
+
+  getText(key) {
+    const MAX_ALIAS_DEPTH = 5;
+    let cursor = { type: 'alias', value: toKey(key) };
+
+    for (
+      let i = 0;
+      i < MAX_ALIAS_DEPTH && cursor && cursor.type === 'alias';
+      i += 1
+    ) {
+      cursor = this._items.get(cursor.value) || null;
+    }
+
+    if (cursor && cursor.type === 'alias') {
+      throw new Error(`Alias depth exceeded when looking up a factoid.`);
+    }
+
+    if (cursor && cursor.type === 'factoid') {
+      return cursor.value;
+    }
+
+    return null;
+  }
+
+  toJSON() {
+    const obj = {};
+    for (const [key, entry] of this._items) {
+      obj[key] = entry;
+    }
+    return obj;
+  }
+
+  needsLoadFromDisk() {
+    return !this._loaded;
+  }
+
+  async loadFromDisk() {
```

Comment:
Need a sanity check here. The goal is to not have two of these running concurrently.